### PR TITLE
Rewrite and improve `ThreadSafeHashtable`

### PR DESCRIPTION
**Pull request by vasilmkd**
I realized that copying the elements of the old array to the beginning of the new array messes up the hashing.

For this one, I dug deep into my Data Structures university course, I think the hash table is finally what it was honestly supposed to be from day 1.

### CODE REVIEW

**Comment by vasilmkd** [comment]
I still have a couple more changes to make.

File: core/shared/src/main/scala/cats/effect/IOFiber.scala [hunk 1]
```diff
@@ -1324,16 +1324,20 @@ private final class IOFiber[A](
     currentCtx.reportFailure(t)
     runtime.shutdown()
 
+    // Make sure the shutdown did not interrupt this thread.
+    Thread.interrupted()
+
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IOFiber.scala:1329]: Sanity-check: do we need to re-interrupt the current thread if we clear the bit here?

Comment (vasilmkd) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala:1329]: What do you mean exactly? I do the interruption at the end of this method.

Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IOFiber.scala:1329]: Oh, gotcha. Okay then!

**Pull request merged**